sources/metricdog/src/main.rs

```diff
@@ -0,0 +1,121 @@
+* `event`: the event that invoked the report.
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `aws_region`: the AWS region the machine is running in.
+* `seed`: the seed value used to roll-out updates.
+* `version_lock`: the optional setting that controls Bottlerocket update selection.
+* `ignore_waves`: an update setting that allows hosts to update before their seed is reached.
+
+### Additionally, when `metricdog` sends a 'health ping', it adds:
+
+* `is_healthy`: true or false based on whether critical services are running.
+* `failed_services`: a list of critical services that have failed, if any.
+
+# Configuration
+
+Configuration is read from a TOML file, which is generated from Bottlerocket settings:
+
+```toml
+# the url to which metricdog will send metrics information
+metrics_url = "https://example.com/metrics"
+# whether or not metricdog will send metrics. opt-out by setting this to false
+send_metrics = true
+# a list of systemd service names that will be checked
+service_checks = ["apiserver", "containerd", "kubelet"]
+# the AWS region
+aws_region = "us-west-2"
+# the update wave seed
+seed = 1234
+# what version bottlerocket should stay on
+version_lock = "latest"
+# whether bottlerocket should ignore update roll-out timing
+ignore_waves = false
+```
+*/
+
+#![deny(rust_2018_idioms)]
+
+mod args;
+mod config;
+mod error;
+mod metricdog;
+#[cfg(test)]
+mod metricdog_test;
+#[cfg(test)]
+mod main_test;
+mod service_check;
+
+use crate::args::{Arguments, Command};
+use crate::config::Config;
+use crate::error::Result;
+use crate::metricdog::Metricdog;
+use crate::service_check::{ServiceCheck, SystemdCheck};
+use bottlerocket_release::BottlerocketRelease;
+use env_logger::Builder as EnvLogger;
+use log::error;
+use snafu::ResultExt;
+use std::process;
+use structopt::StructOpt;
+
+fn main() -> ! {
+    let args = Arguments::from_args();
+    EnvLogger::new()
+        .filter_module("metricdog", args.log_level)
+        .init();
+    process::exit(match main_inner(args, Box::new(SystemdCheck {})) {
+        Ok(()) => 0,
+        Err(err) => {
+            eprintln!("{}", err);
+            1
+        }
+    })
+}
+
+/// pub(crate) for testing.
+pub(crate) fn main_inner(arguments: Arguments, service_check: Box<dyn ServiceCheck>) -> Result<()> {
+    let config = match &arguments.config {
+        None => Config::new()?,
+        Some(filepath) => Config::from_file(filepath)?,
+    };
+    // exit early with no error if the opt-out flag is set
+    if !config.send_metrics {
+        return Ok(());
+    }
+    let os_release = if let Some(os_release_path) = &arguments.os_release {
+        BottlerocketRelease::from_file(os_release_path)
+    } else {
+        BottlerocketRelease::new()
+    }
+    .context(error::BottlerocketRelease)?;
+    let metricdog = Metricdog::from_parts(config, os_release, service_check)?;
+    match arguments.command {
```

Comment:
```suggestion
    .context(error::BottlerocketRelease)?;
    
    let metricdog = Metricdog::from_parts(config, os_release, service_check)?;
    
    match arguments.command {
```
A little breathing room in this function?  :)

Reply:
Aired out in [c966f19](https://github.com/bottlerocket-os/bottlerocket/compare/b8b5c3ad92298db2c46262eb2b6c5b6ea3f60187..c966f19f71c044ce4a20c0a363245035f37b5737)